File: node/tests/key-lookups.test.ts
import { afterEach, describe, expect, it, vi } from 'vitest';
import { HaiClient } from '../src/client.js';
import { generateTestKeypair as generateKeypair } from './setup.js';
import { createMockFFI } from './ffi-mock.js';
import { HaiApiError } from '../src/errors.js';
import { mapFFIError } from '../src/ffi-client.js';

async function makeClient(jacsId: string = 'test-agent'): Promise<HaiClient> {
  const keypair = generateKeypair();
  return HaiClient.fromCredentials(jacsId, keypair.privateKeyPem, { url: 'https://hai.example', privateKeyPassphrase: 'keygen-password' });
}

const KEY_RESPONSE = {
  jacs_id: 'agent-abc',
  version: 'v1',
  public_key: '-----BEGIN PUBLIC KEY-----\nZm9v\n-----END PUBLIC KEY-----\n',
  public_key_raw_b64: 'Zm9v',
  algorithm: 'Ed25519',
  public_key_hash: 'sha256:abcdef1234567890',
  status: 'active',
  dns_verified: true,
  created_at: '2026-01-15T10:30:00Z',
};

const KEY_HISTORY_RESPONSE = {
  jacs_id: 'agent-abc',
  keys: [
    KEY_RESPONSE,
    { ...KEY_RESPONSE, version: 'v0', status: 'rotated' },
  ],
  total: 2,
};

describe('key lookup methods', () => {
  afterEach(() => {
    vi.restoreAllMocks();
  });

  // -----------------------------------------------------------------------
  // fetchKeyByHash
  // -----------------------------------------------------------------------

  describe('fetchKeyByHash', () => {
    it('calls correct URL and parses response', async () => {
      const client = await makeClient();
      const fetchKeyByHashMock = vi.fn(async (hash: string) => {
        expect(hash).toBe('sha256:abcdef1234567890');
        return KEY_RESPONSE;
      });
      client._setFFIAdapter(createMockFFI({ fetchKeyByHash: fetchKeyByHashMock }));

      const result = await client.fetchKeyByHash('sha256:abcdef1234567890');

      expect(result.jacsId).toBe('agent-abc');
      expect(result.algorithm).toBe('Ed25519');
      expect(result.publicKeyHash).toBe('sha256:abcdef1234567890');
      expect(result.dnsVerified).toBe(true);
    });

    it('escapes path-traversal characters in hash', async () => {
      const client = await makeClient();
      const fetchKeyByHashMock = vi.fn(async (hash: string) => {
        // FFI receives raw hash; Rust handles escaping
        expect(hash).toBe('sha256:../../etc/passwd');
        return KEY_RESPONSE;
      });
      client._setFFIAdapter(createMockFFI({ fetchKeyByHash: fetchKeyByHashMock }));

      await client.fetchKeyByHash('sha256:../../etc/passwd');
    });

    it('rejects on 404', async () => {
      const client = await makeClient();
      const fetchKeyByHashMock = vi.fn(async () => {
        throw mapFFIError(new Error('NotFound: key not found'));
      });
      client._setFFIAdapter(createMockFFI({ fetchKeyByHash: fetchKeyByHashMock }));

      await expect(client.fetchKeyByHash('sha256:missing')).rejects.toThrow();
    });
  });

  // -----------------------------------------------------------------------
  // fetchKeyByEmail
  // -----------------------------------------------------------------------

  describe('fetchKeyByEmail', () => {
    it('calls correct URL and parses response', async () => {
      const client = await makeClient();
      const fetchKeyByEmailMock = vi.fn(async (email: string) => {
        expect(email).toBe('alice@hai.ai');
        return KEY_RESPONSE;
      });
      client._setFFIAdapter(createMockFFI({ fetchKeyByEmail: fetchKeyByEmailMock }));

      const result = await client.fetchKeyByEmail('alice@hai.ai');

      expect(result.jacsId).toBe('agent-abc');
      expect(result.version).toBe('v1');
    });

    it('rejects on 404', async () => {
      const client = await makeClient();
      const fetchKeyByEmailMock = vi.fn(async () => {
        throw mapFFIError(new Error('NotFound: key not found'));
      });
      client._setFFIAdapter(createMockFFI({ fetchKeyByEmail: fetchKeyByEmailMock }));

      await expect(client.fetchKeyByEmail('nobody@hai.ai')).rejects.toThrow();
    });
  });

  // -----------------------------------------------------------------------
  // fetchKeyByDomain
  // -----------------------------------------------------------------------

  describe('fetchKeyByDomain', () => {
    it('calls correct URL and parses response', async () => {
      const client = await makeClient();
      const fetchKeyByDomainMock = vi.fn(async (domain: string) => {
        expect(domain).toBe('example.com');
        return KEY_RESPONSE;
      });
      client._setFFIAdapter(createMockFFI({ fetchKeyByDomain: fetchKeyByDomainMock }));

      const result = await client.fetchKeyByDomain('example.com');

      expect(result.jacsId).toBe('agent-abc');
      expect(result.dnsVerified).toBe(true);
    });

    it('rejects on 404', async () => {
      const client = await makeClient();
      const fetchKeyByDomainMock = vi.fn(async () => {
        throw mapFFIError(new Error('NotFound: key not found'));
      });
      client._setFFIAdapter(createMockFFI({ fetchKeyByDomain: fetchKeyByDomainMock }));

      await expect(client.fetchKeyByDomain('nonexistent.test')).rejects.toThrow();
    });
  });

  // -----------------------------------------------------------------------
  // fetchAllKeys
  // -----------------------------------------------------------------------

  describe('fetchAllKeys', () => {
    it('calls correct URL and returns structured result', async () => {
      const client = await makeClient();
      const fetchAllKeysMock = vi.fn(async (jacsId: string) => {
        expect(jacsId).toBe('agent-abc');
        return KEY_HISTORY_RESPONSE;
      });
      client._setFFIAdapter(createMockFFI({ fetchAllKeys: fetchAllKeysMock }));

      const result = await client.fetchAllKeys('agent-abc');

      expect(result.jacsId).toBe('agent-abc');
      expect(result.total).toBe(2);
      expect(result.keys).toHaveLength(2);
      expect(result.keys[0].version).toBe('v1');
      expect(result.keys[1].version).toBe('v0');
    });

    it('escapes jacs_id with slashes', async () => {
      const client = await makeClient();
      const fetchAllKeysMock = vi.fn(async (jacsId: string) => {
        // FFI receives raw jacsId; Rust handles escaping
        expect(jacsId).toBe('agent/with/slashes');
        return KEY_HISTORY_RESPONSE;
      });
      client._setFFIAdapter(createMockFFI({ fetchAllKeys: fetchAllKeysMock }));

      await client.fetchAllKeys('agent/with/slashes');
    });

    it('rejects on 404', async () => {
      const client = await makeClient();
      const fetchAllKeysMock = vi.fn(async () => {
        throw mapFFIError(new Error('NotFound: agent not found'));
      });
      client._setFFIAdapter(createMockFFI({ fetchAllKeys: fetchAllKeysMock }));

      await expect(client.fetchAllKeys('missing-agent')).rejects.toThrow();
    });
  });
});

File: go/keylookups_test.go
package haiai

import (
	"context"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

// keyResponseJSON returns a synthetic key-lookup JSON response.
func keyResponseJSON() string {
	return `{
		"jacs_id":"agent-abc",
		"version":"v1",
		"public_key":"-----BEGIN PUBLIC KEY-----\nZm9v\n-----END PUBLIC KEY-----\n",
		"public_key_raw_b64":"Zm9v",
		"algorithm":"Ed25519",
		"public_key_hash":"sha256:abcdef1234567890",
		"status":"active",
		"dns_verified":true,
		"created_at":"2026-01-15T10:30:00Z"
	}`
}

// keyHistoryResponseJSON returns a synthetic key-history JSON response.
func keyHistoryResponseJSON() string {
	return `{
		"jacs_id":"agent-abc",
		"keys":[` + keyResponseJSON() + `,` + `{
			"jacs_id":"agent-abc",
			"version":"v0",
			"public_key":"-----BEGIN PUBLIC KEY-----\nZm9v\n-----END PUBLIC KEY-----\n",
			"public_key_raw_b64":"Zm9v",
			"algorithm":"Ed25519",
			"public_key_hash":"sha256:abcdef1234567890",
			"status":"rotated",
			"dns_verified":true,
			"created_at":"2025-12-01T10:00:00Z"
		}],
		"total":2
	}`
}

// -----------------------------------------------------------------------
// Deprecated FromURL functions now return errors (no native HTTP fallback)
// -----------------------------------------------------------------------

func TestFetchKeyByEmailFromURL_Deprecated(t *testing.T) {
	_, err := FetchKeyByEmailFromURL(context.Background(), nil, "http://unused", "alice@hai.ai")
	if err == nil {
		t.Fatal("expected error from deprecated FetchKeyByEmailFromURL")
	}
	if !strings.Contains(err.Error(), "deprecated") {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestFetchKeyByDomainFromURL_Deprecated(t *testing.T) {
	_, err := FetchKeyByDomainFromURL(context.Background(), nil, "http://unused", "example.com")
	if err == nil {
		t.Fatal("expected error from deprecated FetchKeyByDomainFromURL")
	}
	if !strings.Contains(err.Error(), "deprecated") {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestFetchAllKeysFromURL_Deprecated(t *testing.T) {
	_, err := FetchAllKeysFromURL(context.Background(), nil, "http://unused", "agent-abc")
	if err == nil {
		t.Fatal("expected error from deprecated FetchAllKeysFromURL")
	}
	if !strings.Contains(err.Error(), "deprecated") {
		t.Fatalf("unexpected error: %v", err)
	}
}

// -----------------------------------------------------------------------
// Standalone functions without client return error
// -----------------------------------------------------------------------

func TestFetchKeyByEmail_NilClient_ReturnsError(t *testing.T) {
	_, err := FetchKeyByEmail(context.Background(), nil, "alice@hai.ai")
	if err == nil {
		t.Fatal("expected error when client is nil")
	}
	if !strings.Contains(err.Error(), "Client required") {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestFetchKeyByDomain_NilClient_ReturnsError(t *testing.T) {
	_, err := FetchKeyByDomain(context.Background(), nil, "example.com")
	if err == nil {
		t.Fatal("expected error when client is nil")
	}
	if !strings.Contains(err.Error(), "Client required") {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestFetchKeyByHash_NilClient_ReturnsError(t *testing.T) {
	_, err := FetchKeyByHash(context.Background(), nil, "sha256:abc")
	if err == nil {
		t.Fatal("expected error when client is nil")
	}
	if !strings.Contains(err.Error(), "Client required") {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestFetchAllKeys_NilClient_ReturnsError(t *testing.T) {
	_, err := FetchAllKeys(context.Background(), nil, "agent-abc")
	if err == nil {
		t.Fatal("expected error when client is nil")
	}
	if !strings.Contains(err.Error(), "Client required") {
		t.Fatalf("unexpected error: %v", err)
	}
}

// -----------------------------------------------------------------------
// Client methods (delegate to FFI via client)
// -----------------------------------------------------------------------

func TestClientFetchKeyByEmail(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		// Go's url.PathEscape does not encode '@' per RFC 3986.
		if r.URL.EscapedPath() != "/api/agents/keys/bob@hai.ai" {
			t.Fatalf("unexpected path: %s", r.URL.EscapedPath())
		}
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(keyResponseJSON()))
	}))
	defer srv.Close()

	t.Setenv("HAI_KEYS_BASE_URL", srv.URL)

	cl, _ := newTestClient(t, srv.URL)
	key, err := cl.FetchKeyByEmail(context.Background(), "bob@hai.ai")
	if err != nil {
		t.Fatalf("FetchKeyByEmail: %v", err)
	}
	if key.AgentID != "agent-abc" {
		t.Fatalf("unexpected agent id: %s", key.AgentID)
	}
}

func TestClientFetchKeyByDomain(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.EscapedPath() != "/api/agents/keys/domain/example.com" {
			t.Fatalf("unexpected path: %s", r.URL.EscapedPath())
		}
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(keyResponseJSON()))
	}))
	defer srv.Close()

	t.Setenv("HAI_KEYS_BASE_URL", srv.URL)

	cl, _ := newTestClient(t, srv.URL)
	key, err := cl.FetchKeyByDomain(context.Background(), "example.com")
	if err != nil {
		t.Fatalf("FetchKeyByDomain: %v", err)
	}
	if key.AgentID != "agent-abc" {
		t.Fatalf("unexpected agent id: %s", key.AgentID)
	}
}

func TestClientFetchAllKeys(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.EscapedPath() != "/api/agents/keys/agent-abc/all" {
			t.Fatalf("unexpected path: %s", r.URL.EscapedPath())
		}
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(keyHistoryResponseJSON()))
	}))
	defer srv.Close()

	t.Setenv("HAI_KEYS_BASE_URL", srv.URL)

	cl, _ := newTestClient(t, srv.URL)
	history, err := cl.FetchAllKeys(context.Background(), "agent-abc")
	if err != nil {
		t.Fatalf("FetchAllKeys: %v", err)
	}
	if history.Total != 2 {
		t.Fatalf("unexpected total: %d", history.Total)
	}
}
